[Time-Series] Autoformer model

elisim opened the pull request:
# What does this PR do?

Adding Time Series Autoformer model https://arxiv.org/abs/2106.13008

Related issue: #21890 

@kashif :) 

## Differences between the vanilla transformer

<img width="484" alt="image" src="https://user-images.githubusercontent.com/17675462/229438415-6a5bba78-c3bf-47b1-966a-f3664a0921e0.png">

* Introduced Series Decomposition in encoder & decoder --- done, waiting for review
* Replaced canonical self-attention with autocorrelation block --- done, waiting for review
* Added seasonal and trend inputs for the decoder --- added todo places in the code
* trend seasonal pseudo code:
```
mean_data = mean(enc_input)
zeros = zeros() # size: x_dec.size(0), prediction_length, x_dec.size(2)
seasonal_init, trend_init = decomp_layer(enc_input)

trend_init = concat(trend_init, mean_data)
seasonal_init = concat(seasonal_init, zeros)
```

elisim committed:
ran `transformers-cli add-new-model-like`

elisim committed:
Merge remote-tracking branch 'upstream/main' into add-autoformer2021-model

elisim committed:
added `AutoformerLayernorm` and `AutoformerSeriesDecomposition`

elisim committed:
added `decomposition_layer` in `init` and `moving_avg` to config

elisim committed:
added `AutoformerAutoCorrelation` to encoder & decoder

elisim committed:
removed caninical self attention `AutoformerAttention`

elisim committed:
added arguments in config and model tester. Init works! 😁

elisim committed:
Merge remote-tracking branch 'upstream/main' into add-autoformer2021-model

elisim committed:
WIP autoformer attention with autocorrlation

elisim committed:
fixed `attn_weights` size

elisim committed:
wip time_delay_agg_training

elisim committed:
Merge remote-tracking branch 'upstream/main' into add-autoformer2021-model

elisim committed:
fixing sizes and debug time_delay_agg_training

elisim committed:
aggregation in training works! 😁

elisim committed:
`top_k_delays` -> `top_k_delays_index` and added `contiguous()`

elisim committed:
wip time_delay_agg_inference

elisim committed:
finish time_delay_agg_inference 😎

elisim committed:
added resize to autocorrelation

elisim committed:
Merge remote-tracking branch 'upstream/main' into add-autoformer2021-model

elisim committed:
bug fix: added the length of the output signal to `irfft`

elisim committed:
`attention_mask = None` in the decoder

elisim committed:
fixed test: changed attention expected size, `test_attention_outputs` works!

elisim committed:
removed unnecessary code

elisim committed:
apply AutoformerLayernorm in final norm in enc & dec

elisim committed:
added series decomposition to the encoder

elisim commented on src/transformers/models/autoformer/modeling_autoformer.py:
```diff
@@ -868,7 +868,7 @@ def forward(
         key_states_fft = torch.fft.rfft(key_states, dim=1)
         attn_weights = query_states_fft * torch.conj(key_states_fft)
-        attn_weights = torch.fft.irfft(attn_weights, dim=1)  # Autocorrelation(Q,K)
+        attn_weights = torch.fft.irfft(attn_weights, n=tgt_len, dim=1)  # Autocorrelation(Q,K)
```
this changed follows
https://github.com/thuml/Autoformer/issues/133

elisim committed:
Merge remote-tracking branch 'upstream/main' into add-autoformer2021-model

elisim committed:
added series decomp to decoder, with inputs

elisim committed:
added trend todos

elisim committed:
added autoformer to README

elisim committed:
added to index

elisim committed:
added autoformer.mdx

elisim committed:
remove scaling and init attention_mask in the decoder

elisim committed:
make style

elisim committed:
fix copies

elisim committed:
make fix-copies

elisim commented on src/transformers/models/autoformer/modeling_autoformer.py:
```diff
@@ -1570,20 +1572,23 @@ class AutoformerModel(AutoformerPreTrainedModel):
         # transformer encoder-decoder and mask initializer
         self.encoder = AutoformerEncoder(config)
         self.decoder = AutoformerDecoder(config)
-        self.decomposition_layer = AutoformerSeriesDecompositionLayer(config.moving_avg)
```
revert this delete after Informer merge

elisim replied:
added in https://github.com/huggingface/transformers/pull/21891/commits/e61dac5b68baefdcbd339de545041e55ae535131

elisim commented on src/transformers/models/autoformer/modeling_autoformer.py:
```diff
@@ -1802,11 +1793,7 @@ def forward(
 
-        # decoder input
         dec_input = transformer_inputs[:, self.config.context_length :, ...]
-        # mean_enc_input = enc_input.mean(dim=1).unsqueeze(1).repeat(1, dec_input.size(1), 1)
```
same, revert this delete after Informer merge

elisim replied:
done in https://github.com/huggingface/transformers/pull/21891/commits/bbb722c0b3ebe8be4b3725e206d8b1bd11bee312

elisim commented:
One small open issue left, is adding the series decomposition to the decoder with the trend input. Will do after the initial review :)

kashif committed:
Merge remote-tracking branch 'upstream/main' into add-autoformer2021-model

kashif committed:
inital fix-copies

elisim committed:
Merge remote-tracking branch 'upstream/main' into add-autoformer2021-model

elisim committed:
Merge branch 'add-autoformer2021-model' of https://github.com/elisim/transformers into add-autoformer2021-model

elisim committed:
fix from https://github.com/huggingface/transformers/pull/22076

kashif committed:
Merge remote-tracking branch 'upstream/main' into add-autoformer2021-model

kashif committed:
make style

kashif committed:
fix class names

kashif committed:
Merge remote-tracking branch 'elisim/main' into add-autoformer2021-model

kashif committed:
added trend

kashif committed:
added d_model and projection layers

elisim committed:
added `trend_projection` source, and decomp layer init

elisim committed:
Merge branch 'main' into add-autoformer2021-model

elisim committed:
added trend & seasonal init for decoder input

kashif committed:
AutoformerModel cannot be copied as it has the decomp layer too

kashif committed:
encoder can be copied from time series transformer

kashif committed:
fixed generation and made distrb. out more robust

kashif committed:
use context window to calculate decomposition

kashif committed:
use the context_window for decomposition

kashif committed:
use output_params helper

kashif committed:
clean up AutoformerAttention

kashif committed:
Merge branch 'huggingface:main' into add-autoformer2021-model

elisim commented on src/transformers/models/autoformer/modeling_autoformer.py:
```diff
@@ -522,10 +522,10 @@ def forward(
         # find top k autocorrelations delays
         top_k = int(self.factor * math.log(time_length))
-        autocorrelations_mean_on_head_channel = torch.mean(torch.mean(autocorrelations, dim=1), dim=2)  # bsz x tgt_len
+        autocorrelations_mean_on_head_channel = torch.mean(autocorrelations, dim=(1, -1))  # bsz x tgt_len
```
Good!

elisim commented on src/transformers/models/autoformer/modeling_autoformer.py:
```diff
@@ -522,10 +522,10 @@ def forward(
         if self.training:
             autocorrelations_mean_on_bsz = torch.mean(autocorrelations_mean_on_head_channel, dim=0)
-            top_k_delays_index = torch.topk(autocorrelations_mean_on_bsz, top_k)[1]
+            _, top_k_delays_index = torch.topk(autocorrelations_mean_on_bsz, top_k)
```
thanks for the good improvements!

kashif committed:
subsequences_length off by 1

kashif committed:
make fix copies

kashif committed:
fix test

elisim commented on tests/models/autoformer/test_modeling_autoformer.py:
```diff
@@ -297,19 +302,17 @@ def test_attention_outputs(self):
-                outputs = model(**inputs)
+                outputs = model(**self._prepare_for_class(inputs_dict, model_class))
             attentions = outputs.encoder_attentions
-            channel = inputs["past_time_features"].size(-1)
```
Ohhh using the dim param is actually better! Will know for the future

elisim committed:
added init for nn.Conv1d

kashif committed:
fix IGNORE_NON_TESTED

kashif committed:
added model_doc

kashif committed:
fix ruff

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_21891). All of your documentation changes will be reflected on that endpoint.

kashif committed:
ignore tests

kashif committed:
remove dup

kashif committed:
fix SPECIAL_CASES_TO_ALLOW

kashif committed:
do not copy due to conv1d weight init

kashif committed:
remove unused imports

kashif committed:
added short summary

kashif committed:
added label_length and made the model non-autoregressive

elisim commented on src/transformers/models/autoformer/configuration_autoformer.py:
```diff
@@ -162,13 +162,15 @@ def __init__(
         init_std: float = 0.02,
         use_cache=True,
         # Autoformer arguments
+        label_length: Optional[int] = None,
```
This is great. I'm thinking after merging Autoformer, maybe it's possibe to add the non-autoregressive generation to Informer & vanilla also? @kashif

elisim replied:
Talked in Slack

elisim reviewed (commented):
added thoughts about non-autoregressive generation

elisim committed:
added params docs

elisim committed:
better doc for `factor`

kashif committed:
fix tests

elisim committed:
Merge remote-tracking branch 'upstream/main' into add-autoformer2021-model

elisim committed:
renamed `moving_avg` to `moving_average`

elisim committed:
renamed `factor` to `autocorrelation_factor`

elisim committed:
make style

NielsRogge commented on src/transformers/models/autoformer/configuration_autoformer.py:
```diff
@@ -0,0 +1,249 @@
+    This is the configuration class to store the configuration of a [`AutoformerModel`]. It is used to instantiate a
+    Autoformer model according to the specified arguments, defining the model architecture. Instantiating a
+    configuration with the defaults will yield a similar configuration to that of the Time Series Transformer
+    [huggingface/time-series-transformer-tourism-monthly](https://huggingface.co/huggingface/time-series-transformer-tourism-monthly)
```
to be updated

NielsRogge commented on src/transformers/models/autoformer/configuration_autoformer.py:
```diff
@@ -0,0 +1,249 @@
+
+class AutoformerConfig(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a [`AutoformerModel`]. It is used to instantiate a
+    Autoformer model according to the specified arguments, defining the model architecture. Instantiating a
```
```suggestion
    This is the configuration class to store the configuration of an [`AutoformerModel`]. It is used to instantiate an
    Autoformer model according to the specified arguments, defining the model architecture. Instantiating a
```

NielsRogge commented on src/transformers/models/autoformer/configuration_autoformer.py:
```diff
@@ -0,0 +1,249 @@
+        self.output_attentions = False
+        self.output_hidden_states = False
+
+        self.use_cache = use_cache
```
The `use_cache` argument can be removed I think, since this is not used. It can be used to speed up autoregressive generation by caching the precomputed keys and values

NielsRogge commented on src/transformers/models/autoformer/configuration_autoformer.py:
```diff
@@ -0,0 +1,249 @@
+        self.activation_function = activation_function
+        self.init_std = init_std
+
+        self.output_attentions = False
+        self.output_hidden_states = False
```
These arguments shouldn't be added normally

elisim replied:
Good spot. I checked in the previous models, and these arguments are set their too:

https://github.com/huggingface/transformers/blob/312b104ff65514736c0475814fec19e47425b0b5/src/transformers/models/time_series_transformer/configuration_time_series_transformer.py#L220

https://github.com/huggingface/transformers/blob/312b104ff65514736c0475814fec19e47425b0b5/src/transformers/models/informer/configuration_informer.py#L235

Is it a mistake? or should I remove them also?

kashif replied:
removed

elisim committed:
Update src/transformers/models/autoformer/configuration_autoformer.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

NielsRogge commented on src/transformers/models/autoformer/configuration_autoformer.py:
```diff
@@ -0,0 +1,249 @@
+# coding=utf-8
+# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
```suggestion
```
(nit) Typically only attribution is given in the modeling file as the config is HF specific

NielsRogge commented on src/transformers/models/autoformer/modeling_autoformer.py:
```diff
@@ -0,0 +1,2008 @@
+        ... )
+        >>> batch = torch.load(file)
+
+        >>> model = AutoformerForPrediction.from_pretrained("huggingface/time-series-transformer-tourism-monthly")
```
To be updated

elisim committed:
Update src/transformers/models/autoformer/configuration_autoformer.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

kashif committed:
fix configurations

kashif committed:
fix integration tests

elisim committed:
Merge remote-tracking branch 'upstream/main' into add-autoformer2021-model

elisim committed:
Merge remote-tracking branch 'upstream/main' into add-autoformer2021-model

elisim committed:
Merge remote-tracking branch 'upstream/main' into add-autoformer2021-model

elisim commented:
PR is green

amyeroberts reviewed (commented):
Thanks for adding this model! 

Overall nice and clean PR, just a few comments on a few additional pieces to be implemented: 
* The shapes of the tensors weren't always clear to me based on the model outputs classes and docstrings. As I've understood, the output is quite specific for this model and so everything should match and be made explicit particularly when it deviates from the library defaults. 
* All model layers should take the config in its init and use that to initialise its values. 
* Test methods should be overridden with model-specific logic if defaults aren't compatible, and only skipped if not applicable. 
* There's parts of the code incl. docstrings which I believe are copy-pasted from time series transformers. Please make sure everything has been adapted for this specific model e.g. I tried running the example in the `AutoformerForPrediction` docstring and it failed with an error which I suspect is because of this.

kashif commented:
thank you @amyeroberts will get it fixed!

elisim commented:
@amyeroberts, thank you for the comprehensive CR! I sincerely appreciate the effort and time you dedicated to thoroughly assessing this pull request. 

Will be fixed!

kashif committed:
Update src/transformers/models/autoformer/configuration_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

kashif commented on src/transformers/models/autoformer/configuration_autoformer.py:
```diff
@@ -0,0 +1,245 @@
+
+    def __init__(
+        self,
+        prediction_length: Optional[int] = None,
```
yes it must be set but its dataset specific so it is not clear what the default should be...

amyeroberts replied:
OK - in this case could we raise an exception in the config init if not set?

kashif replied:
we raise an exception in the model when it is not set

kashif replied:
i believe the main issue was that if we put the validations in the configuratoin initializers then a lot of the common tests for the config were failing...

elisim commented on utils/check_repo.py:
```diff
@@ -73,6 +73,8 @@
     "TimeSeriesTransformerDecoder",  # Building part of bigger (tested) model.
     "InformerEncoder",  # Building part of bigger (tested) model.
     "InformerDecoder",  # Building part of bigger (tested) model.
+    "AutoformerEncoder",  # Building part of bigger (tested) model.
```
Currently it's grouped with the other time-series models:

```
"TimeSeriesTransformerEncoder",  # Building part of bigger (tested) model.
"TimeSeriesTransformerDecoder",  # Building part of bigger (tested) model.
"InformerEncoder",  # Building part of bigger (tested) model.
"InformerDecoder",  # Building part of bigger (tested) model.
"AutoformerEncoder",  # Building part of bigger (tested) model.
"AutoformerDecoder",  # Building part of bigger (tested) model.
```

Re-top it will break this convention, what do you think?

amyeroberts replied:
Makes sense :) 👍

elisim commented on utils/check_repo.py:
```diff
@@ -223,6 +225,7 @@
     "GPTSanJapaneseModel",
     "TimeSeriesTransformerForPrediction",
     "InformerForPrediction",
+    "AutoformerForPrediction",
```
Same as before

elisim committed:
fixing `lags_sequence` doc

elisim commented on tests/models/autoformer/test_modeling_autoformer.py:
```diff
@@ -0,0 +1,482 @@
+
+        expected_slice = torch.tensor([3130.6763, 4056.5293, 7053.0786], device=torch_device)
+        mean_prediction = outputs.sequences.mean(dim=1)
+        self.assertTrue(torch.allclose(mean_prediction[0, -3:], expected_slice, rtol=1e-1))
```
It was generated from the original time-series-transformer (when I ran `transformers-cli add-new-model-like`). 
As can be seen in Informer and TimeSeriesTransformer:
https://github.com/huggingface/transformers/blob/3658488ff77ff8d45101293e749263acf437f4d5/tests/models/time_series_transformer/test_modeling_time_series_transformer.py#L536

https://github.com/huggingface/transformers/blob/3658488ff77ff8d45101293e749263acf437f4d5/tests/models/informer/test_modeling_informer.py#L513

cc @kashif

kashif replied:
so this is relative tolerance `rtol` rather than absolute which is used since the values for the predictions are huge and this makes it tight too and a bit more robust

elisim commented on src/transformers/models/autoformer/modeling_autoformer.py:
```diff
@@ -0,0 +1,2010 @@
+
+
+AUTOFORMER_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "thuml/Autoformer",
```
good catch! :) The original repo https://github.com/thuml/Autoformer writes it as `thuml`

elisim commented on src/transformers/models/autoformer/configuration_autoformer.py:
```diff
@@ -0,0 +1,245 @@
+            Whether to scale the input targets.
+        lags_sequence (`list[int]`, *optional*, defaults to `[1, 2, 3, 4, 5, 6, 7]`):
+            The lags of the input time series as covariates often dictated by the frequency. Default is `[1, 2, 3, 4,
+            5, 6, 7]`.
```
It's actually done by the autoformatting. I tried to change it in this commit https://github.com/huggingface/transformers/pull/21891/commits/21e34911e36a6f8f45f25cbf43584a49e5316c55 but the code-quality test is falling because of that

elisim committed:
Revert "fixing `lags_sequence` doc"

This reverts commit 21e34911e36a6f8f45f25cbf43584a49e5316c55.

elisim committed:
Update src/transformers/models/autoformer/modeling_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

elisim committed:
Update src/transformers/models/autoformer/modeling_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

elisim committed:
Update src/transformers/models/autoformer/modeling_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

elisim committed:
Apply suggestions from code review

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

elisim commented on src/transformers/models/informer/configuration_informer.py:
```diff
@@ -232,9 +232,6 @@ def __init__(
         self.activation_function = activation_function
         self.init_std = init_std
 
-        self.output_attentions = False
```
cc @kashif

elisim committed:
Update src/transformers/models/autoformer/configuration_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

elisim committed:
model layers now take the config

elisim commented on src/transformers/models/autoformer/modeling_autoformer.py:
```diff
@@ -0,0 +1,2010 @@
+# Class based on
+# https://github.com/thuml/Autoformer/blob/c6a0694ff484753f2d986cc0bb1f99ee850fc1a8/layers/Autoformer_EncDec.py#L39
+# where AutoformerSeriesDecompositionLayer is series_decomp + moving_average
+class AutoformerSeriesDecompositionLayer(nn.Module):
```
done in https://github.com/huggingface/transformers/pull/21891/commits/6bfe3d2a776c17bb4c960e2fc0a13d202ee7cab0

elisim commented on src/transformers/models/autoformer/modeling_autoformer.py:
```diff
@@ -0,0 +1,2010 @@
+    - torch.mean(nn.LayerNorm(x))
+    """
+
+    def __init__(self, channels):
```
done in https://github.com/huggingface/transformers/pull/21891/commits/6bfe3d2a776c17bb4c960e2fc0a13d202ee7cab0

elisim committed:
added `layer_norm_eps` to the config

elisim commented on src/transformers/models/autoformer/modeling_autoformer.py:
```diff
@@ -0,0 +1,2010 @@
+
+    def __init__(self, channels):
+        super().__init__()
+        self.layernorm = nn.LayerNorm(channels)
```
added `layer_norm_eps` to the config in https://github.com/huggingface/transformers/pull/21891/commits/a93498c03494f3f9bd901a366ac50985584bf870

elisim committed:
Update src/transformers/models/autoformer/modeling_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

elisim committed:
added `config.layer_norm_eps` to AutoformerLayernorm

elisim committed:
added `config.layer_norm_eps` to all layernorm layers

elisim committed:
Update src/transformers/models/autoformer/configuration_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

elisim committed:
Update src/transformers/models/autoformer/configuration_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

elisim committed:
Update src/transformers/models/autoformer/configuration_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

elisim committed:
Update src/transformers/models/autoformer/configuration_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

elisim commented on src/transformers/models/autoformer/modeling_autoformer.py:
```diff
@@ -0,0 +1,2010 @@
+        )
+        hidden_states = nn.functional.dropout(hidden_states, p=self.dropout, training=self.training)
+        hidden_states = residual + hidden_states
+        # added layer norm here as an improvement
```
Original impl removed the layernorm layers here. We decided to keep them 🙂

amyeroberts replied:
OK - I'm actually not sure what the standard for the library here would be: whether the original architecture should take precedence or not. Pinging @sgugger who will know :)

elisim commented on src/transformers/models/autoformer/modeling_autoformer.py:
```diff
@@ -0,0 +1,2010 @@
+        hidden_states = nn.functional.dropout(hidden_states, p=self.dropout, training=self.training)
+        hidden_states = residual + hidden_states
+        hidden_states, trend1 = self.decomp1(hidden_states)
+        # added layer norm here as an improvement
```
Original impl removed the layernorm layers here. We decided to keep them 🙂

elisim committed:
fix variable names

elisim commented:
* Addressed questions I could answer
* Added `layer_norm_eps` 
* Model layers are now take the config, except the `AutoformerAttention` which I wasn't sure about

kashif committed:
added inital pretrained model

kashif committed:
added use_cache docstring

kashif committed:
doc strings for trend and use_cache

kashif committed:
fix order of args

kashif committed:
imports on one line

kashif committed:
fixed get_lagged_subsequences docs

kashif committed:
add docstring for create_network_inputs

kashif commented on src/transformers/models/autoformer/modeling_autoformer.py:
```diff
@@ -0,0 +1,2010 @@
+    This block replace the canonical self-attention mechanism.
+    """
+
+    def __init__(
+        self,
+        embed_dim: int,
+        num_heads: int,
+        dropout: float = 0.0,
+        is_decoder: bool = False,
+        bias: bool = True,
+        autocorrelation_factor: int = 3,
```
so we took inspiration from the `BartAttention` module here as some of the arguments are not in the config I belive like bias etc and other arguments are different for the encoder or decoder part. Thus the appropriate encoder/decoder layer uses the `config` to initialize this module...

kashif committed:
get rid of layer_norm_eps config

elisim commented on src/transformers/models/autoformer/configuration_autoformer.py:
```diff
@@ -106,8 +106,6 @@ class AutoformerConfig(PretrainedConfig):
             The standard deviation of the truncated normal weight initialization distribution.
         use_cache (`bool`, *optional*, defaults to `True`):
             Whether to use the past key/values attentions (if applicable to the model) to speed up decoding.
-        layer_norm_eps (`float`, *optional*, defaults to 1e-6):
```
Why removing the layer norm eps? It was added because the CR review

elisim reviewed (commented):
Explanation for the layer norm deletion

kashif committed:
add back layernorm

kashif committed:
update fixture location

kashif committed:
fix signature

kashif committed:
use AutoformerModelOutput dataclass

kashif commented on src/transformers/models/autoformer/modeling_autoformer.py:
```diff
@@ -0,0 +1,2010 @@
+            outputs = ((params,) + outputs[1:]) if params is not None else outputs[1:]
+            return ((prediction_loss,) + outputs) if prediction_loss is not None else outputs
+
+        return Seq2SeqTSPredictionOutput(
```
added `AutoformerModelOutput` and fixed some of the tests

amyeroberts commented on src/transformers/models/autoformer/modeling_autoformer.py:
```diff
@@ -0,0 +1,2062 @@
+
+
+@dataclass
+class AutoformerModelOutput(Seq2SeqTSModelOutput):
```
Rather than subclass `Seq2SeqTSModelOutput`, this should subclass `ModelOutput` and implement all of the attributes directly. 

It's true that there's some repetition but is more inline with the one-model one-file philosophy and makes sure the docstring is accurate.

amyeroberts commented on src/transformers/models/autoformer/configuration_autoformer.py:
```diff
@@ -0,0 +1,249 @@
+        self.layer_norm_eps = layer_norm_eps
+
+        # Autoformer
+        self.label_length = label_length or 10
```
```suggestion
        self.label_length = label_length
```

kashif replied:
@amyeroberts so if the configs do not have a default value the common tests for the config fail... i believe in transformers one needs a default value for all arguments...

amyeroberts replied:
Right, but isn't it already set as a default value as a kwarg above?

amyeroberts commented on tests/models/autoformer/test_modeling_autoformer.py:
```diff
@@ -0,0 +1,469 @@
+    def test_hidden_states_output(self):
+        pass
+
+    @unittest.skip(reason="Result of the model has randomness")
+    def test_model_outputs_equivalence(self):
+        pass
```
As this test is checking the equivalence when `return_dict=True` and `return_dict=False` - couldn't we set a seed for the two forward passes and compare?

amyeroberts commented on tests/models/autoformer/test_modeling_autoformer.py:
```diff
@@ -0,0 +1,469 @@
+    def test_resize_tokens_embeddings(self):
+        pass
+
+    @unittest.skip(reason="Result of the decoder's hidden state is a tuple")
```
It should still be tested :)

amyeroberts reviewed (commented):
Looking good - thanks again for the work adding this model! 

There's just a few small things to address before we're ready to merge: 
* Validation of config arguments in the config class 
* The dataclass for the specific model output should subclass from `ModelOutput` and implement all of it's attributes explicitly
* Some tests are still skipped and should be tested
* There's different epsilon values used to instantiate layer norms throughout the model. If different layers have different values, then we should handle this in the config.

kashif committed:
fix pretrain config

kashif commented on src/transformers/models/autoformer/configuration_autoformer.py:
```diff
@@ -0,0 +1,249 @@
+            The standard deviation of the truncated normal weight initialization distribution.
+        use_cache (`bool`, *optional*, defaults to `True`):
+            Whether to use the past key/values attentions (if applicable to the model) to speed up decoding.
+        layer_norm_eps (`float`, *optional*, defaults to 1e-6):
```
right we can remove this config and keep it to the default one from pytorch.... i do not think it makes sense to have an extra layer_norm_eps here

kashif committed:
no need as default exists

kashif committed:
subclass ModelOutput

kashif committed:
remove layer_norm_eps config

kashif committed:
fix test_model_outputs_equivalence test

kashif committed:
test hidden_states_output

kashif committed:
make fix-copies

amyeroberts reviewed (approved):
Thanks for adding this model!

Mostly just a few remaining nits. Main comment is about some of the testing logic. Then I think we're good to merge :)

elisim committed:
Update src/transformers/models/autoformer/configuration_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

kashif commented on tests/models/autoformer/test_modeling_autoformer.py:
```diff
@@ -0,0 +1,461 @@
+
+        seq_len = getattr(self.model_tester, "seq_length", None)
+        decoder_seq_length = getattr(self.model_tester, "decoder_seq_length", seq_len)
+        getattr(self.model_tester, "label_length", seq_len)
```
removed!

kashif committed:
removed unused attr

kashif committed:
Update tests/models/autoformer/test_modeling_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

kashif committed:
Update src/transformers/models/autoformer/modeling_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

kashif committed:
Update src/transformers/models/autoformer/modeling_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

kashif committed:
Update src/transformers/models/autoformer/modeling_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

kashif committed:
Update src/transformers/models/autoformer/modeling_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

kashif committed:
Update src/transformers/models/autoformer/modeling_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

kashif committed:
Update src/transformers/models/autoformer/modeling_autoformer.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

kashif committed:
use AutoFormerDecoderOutput

kashif committed:
fix formatting

kashif commented on src/transformers/models/autoformer/modeling_autoformer.py:
```diff
@@ -0,0 +1,2177 @@
+        """
+        Returns lagged subsequences of a given sequence. Returns a tensor of shape (batch_size, subsequences_length,
+        feature_size, indices_length),
+            containing lagged subsequences. Specifically, lagged[i, j, :, k] = sequence[i, -indices[k]-S+j, :].
```
fixed its `subsequences_length`

kashif committed:
fix formatting
